crates/astria-conductor/src/data_availability.rs

```diff
@@ -0,0 +1,382 @@
+        block_verifier: BlockVerifier,
+        namespace: Namespace,
+        shutdown: oneshot::Receiver<()>,
+    ) -> eyre::Result<Self> {
+        let celestia_client = CelestiaClient::builder()
+            .endpoint(celestia_node_url)
+            .bearer_token(celestia_bearer_token)
+            .build()
+            .wrap_err("failed creating celestia client")?;
+
+        // TODO: we should probably pass in the height we want to start at from some genesis/config
+        // file
+        let current_block_height = celestia_client.get_latest_height().await?;
+        info!(da_height = current_block_height, "creating Reader");
+
+        Ok(Self {
+            executor_tx,
+            celestia_client,
+            celestia_poll_interval,
+            current_block_height,
+            get_latest_height: None,
+            get_sequencer_datas: JoinMap::new(),
+            process_sequencer_datas: JoinMap::new(),
+            block_verifier,
+            namespace,
+            shutdown,
+        })
+    }
+
+    #[instrument(skip(self))]
+    pub(crate) async fn run_until_stopped(mut self) -> eyre::Result<()> {
+        info!("Starting reader event loop.");
+
+        let mut interval = tokio::time::interval(self.celestia_poll_interval);
+        loop {
+            select!(
+                shutdown_res = &mut self.shutdown => {
+                    match shutdown_res {
+                        Ok(()) => info!("received shutdown command; exiting"),
+                        Err(e) => warn!(error.message = %e, error.cause = ?e, "shutdown receiver dropped; exiting"),
+                    }
+                    break;
+                }
+
+                _ = interval.tick() => self.get_latest_height(),
+
+                res = async { self.get_latest_height.as_mut().unwrap().await }, if self.get_latest_height.is_some() => {
+                    self.get_latest_height = None;
+                    self.get_sequencer_datas(res);
+                }
+
+                Some((height, res)) = self.get_sequencer_datas.join_next(), if !self.get_sequencer_datas.is_empty() => {
+                    self.process_sequencer_datas(height, res);
+                }
+
+                Some((height, res)) = self.process_sequencer_datas.join_next(), if !self.process_sequencer_datas.is_empty() => {
+                    let span = tracing::info_span!("send_sequencer_subsets", height);
+                    span.in_scope(|| self.send_sequencer_subsets(res))
+                        .wrap_err("failed sending sequencer subsets to executor")?;
+                }
+            )
+        }
+        Ok(())
+    }
+
+    fn get_latest_height(&mut self) {
+        let client = self.celestia_client.inner().clone();
+        self.get_latest_height = Some(tokio::spawn(async move {
+            Ok(client.header_network_head().await?.height())
+        }))
+    }
+
+    fn get_sequencer_datas(&mut self, latest_height_res: Result<eyre::Result<u64>, JoinError>) {
```

Comment:
I think this function name is a bit confusing, it matches `get_sequencer_datas` `JoinSet` but it only sometimes does things related, and then only instantiates jobs on that set?

When reading from top I was concerned based on naming we were double fetching, maybe something like `start_data_availability_fetch`?

Reply:
can you add a comment explaining `latest_height_res` is the celestia height you're fetching at

Reply:
I have renamed the function to `fetch_sequencer_blobs_up_to_latest_height, added a comment, and renamed the joinmap to `fetch_sequencer_blobs_at_height`.